sparse/client.go

```diff
@@ -139,334 +105,233 @@ func connect(host, port string, timeout int) net.Conn {
-				processFileInterval(subrange, rrange, netStream)
-				if len(data) > 0 {
-					lrange.Data = lrange.Data[subrange.Len():]
-				}
-				lrange.Begin = rrange.End
-				rrange = <-remote
-				continue
-			} else if lrange.End < rrange.End {
-				if verboseClient {
-					logData("LHASH", lrange.Data)
+		// syncing data interval
+		err := session.SyncDataInterval(file, interval)
+		if err != nil {
+			log.Debugf("SyncDataInterval failed: %s", interval)
+			return err
+		}
+		if e.Flags&FIEMAP_EXTENT_LAST != 0 {
+			log.Debugf("hit FIEMAP_EXTENT_LAST flag, are we on last index yet ? %v", (index == len(exts)-1))
+
+			// report last hole
+			if lastIntervalEnd < fileSize {
+				holeInterval := Interval{lastIntervalEnd, fileSize}
+				log.Debugf("Here is a hole: %s", holeInterval)
+
+				// syncing hole interval
+				err = session.SyncHoleInterval(holeInterval)
+				if err != nil {
+					log.Debugf("SyncHoleInterval failed: %s", holeInterval)
+					return err
 				}
-
-				hashFileData(fileHasher, lrange.Len(), lrange.Data)
-				processFileInterval(lrange, HashedInterval{FileInterval{rrange.Kind, lrange.Interval}, make([]byte, 0)}, netStream)
-				rrange.Begin = lrange.End
-				lrange = <-local
-				continue
 			}
-			if verboseClient {
-				logData("LHASH", lrange.Data)
-			}
-			hashFileData(fileHasher, lrange.Len(), lrange.Data)
-			processFileInterval(lrange, rrange, netStream)
-			lrange = <-local
-			rrange = <-remote
-		} else {
-			// Should never happen
-			log.Fatal("processDiff internal error")
-			return
 		}
 	}
-	log.Info("Finished processing file diff")
 
-	status := true
-	err = <-errStream
-	if err != nil {
-		log.Error("Sync client file load aborted:", err)
-		status = false
-	}
-	// make sure we finished consuming dst hashes
-	status = <-netInStreamDone && status // netDstReceiver finished
-	log.Info("Finished consuming remote file hashes, status=", status)
-
-	// Send end of transmission
-	netStream <- diffChunk{true, DataInterval{FileInterval{SparseIgnore, Interval{0, 0}}, make([]byte, 0)}}
-
-	// get network sender status
-	net := <-netStatus
-	log.Info("Finished sending file diff of", net.byteCount, "(bytes), status=", net.status)
-	if !net.status {
-		err = errors.New("netwoek transfer failure")
-		return
+	// special case, the whole file is a hole
+	if len(exts) == 0 && fileSize != 0 {
+		// report hole
+		holeInterval = Interval{0, fileSize}
+		log.Debugf("The file is a hole: %s", holeInterval)
+
+		// syncing hole interval
+		err := session.SyncHoleInterval(holeInterval)
+		if err != nil {
+			log.Debugf("SyncHoleInterval failed: %s", holeInterval)
+			return err
+		}
 	}
 
-	var statusRemote bool
-	err = decoder.Decode(&statusRemote)
+	session.encoder.Encode(requestHeader{requestMagic, syncDone})
+	return nil
+}
+
+func (session *SyncSessionClient) sendSyncRequest(path string, size int64) error {
+	err := session.encoder.Encode(requestHeader{requestMagic, syncRequestCode})
```

Comment:
I'd rather create a method in `session`, for encode/decode. 

Also, it's misleading name here because what's really happening is encode and send etc.

So it's better to be called `session.Send()` and `session.Receive()`